include/pybind11/pybind11.h

```diff
@@ -1374,6 +1374,141 @@ using default_holder_type = smart_holder;
+// This implementation works as-is for smart_holder std::shared_ptr members.
+template <typename T, typename D, typename SFINAE = void>
+struct xetter_cpp_function {
+    template <typename PM>
+    static cpp_function readonly(PM pm, const handle &hdl) {
+        return cpp_function([pm](const T &c) -> const D & { return c.*pm; }, is_method(hdl));
+    }
+
+    template <typename PM>
+    static cpp_function read(PM pm, const handle &hdl) {
+        return readonly(pm, hdl);
+    }
+
+    template <typename PM>
+    static cpp_function write(PM pm, const handle &hdl) {
+        return cpp_function([pm](T &c, const D &value) { c.*pm = value; }, is_method(hdl));
+    }
+};
+
+// smart_holder specializations for raw pointer members.
+// WARNING: Like the classic implementation, this implementation can lead to dangling pointers.
+// See test_ptr() in tests/test_class_sh_property.py
+// However, the read functions return a shared_ptr to the member, emulating PyCLIF approach:
+// https://github.com/google/clif/blob/c371a6d4b28d25d53a16e6d2a6d97305fb1be25a/clif/python/instance.h#L233
+// This prevents disowning of the Python object owning the raw pointer member.
+template <typename T, typename D>
+struct xetter_cpp_function<
+    T,
+    D,
+    detail::enable_if_t<detail::type_uses_smart_holder_type_caster<T>::value
+                        && detail::type_uses_smart_holder_type_caster<D>::value
+                        && std::is_pointer<D>::value>> {
+
+    using drp = typename std::remove_pointer<D>::type;
+
+    template <typename PM>
+    static cpp_function readonly(PM pm, const handle &hdl) {
+        return cpp_function(
+            [pm](const std::shared_ptr<T> &c_sp) -> std::shared_ptr<drp> {
+                D ptr = (*c_sp).*pm;
+                return std::shared_ptr<drp>(c_sp, ptr);
+            },
+            is_method(hdl));
+    }
+
+    template <typename PM>
+    static cpp_function read(PM pm, const handle &hdl) {
+        return readonly(pm, hdl);
+    }
+
+    template <typename PM>
+    static cpp_function write(PM pm, const handle &hdl) {
+        return cpp_function([pm](T &c, D value) { c.*pm = value; }, is_method(hdl));
+    }
+};
+
+// smart_holder specializations for members held by-value.
+// The read functions return a shared_ptr to the member, emulating PyCLIF approach:
+// https://github.com/google/clif/blob/c371a6d4b28d25d53a16e6d2a6d97305fb1be25a/clif/python/instance.h#L233
+// This prevents disowning of the Python object owning the member.
+template <typename T, typename D>
+struct xetter_cpp_function<
+    T,
+    D,
+    detail::enable_if_t<detail::type_uses_smart_holder_type_caster<T>::value
+                        && detail::type_uses_smart_holder_type_caster<D>::value
+                        && !std::is_pointer<D>::value           //
+                        && !detail::is_std_unique_ptr<D>::value //
+                        && !detail::is_std_shared_ptr<D>::value>> {
```

Comment:
```suggestion
    detail::enable_if_t<detail::all_of<
                        detail::type_uses_smart_holder_type_caster<T>,
                        detail::type_uses_smart_holder_type_caster<D>,
                        !std::is_pointer<D>,           //
                        !detail::is_std_unique_ptr<D>, //
                        !detail::is_std_shared_ptr<D>>>::value> {

Reply:
Thanks for pointing this out! It looks a lot nicer.